File: os-compass/src/api/category.ts
import { invoke } from '@tauri-apps/api/core';

export interface Category {
  id: number;
  name: string;
  path: string;
  explain?: string;
  sort_order: number;
  parent_id?: number;
  is_system: boolean;
  project_count: number;
  created_at: string;
}

export interface CreateCategoryInput {
  name: string;
  path: string;
  explain?: string;
  sort_order?: number;
  parent_id?: number;
}

export interface PreviewItem {
  full_path: string;
  status: "new" | "skip" | "error";
  reason?: string | null;
  source_line: number;
}

export interface ParsedCategories {
  total: number;
  items: PreviewItem[];
}

export interface ImportError {
  source_line: number;
  full_path: string;
  reason: string;
}

export interface ImportSummary {
  created: number;
  skipped: number;
  errors: ImportError[];
}

export const categoryApi = {
  async getAll(): Promise<Category[]> {
    return invoke('get_categories');
  },

  async getById(id: number): Promise<Category | null> {
    return invoke('get_category', { id });
  },

  async create(input: CreateCategoryInput): Promise<number> {
    return invoke('create_category', { input });
  },

  async update(category: Category): Promise<void> {
    return invoke('update_category', { category });
  },

  async delete(id: number): Promise<void> {
    return invoke('delete_category', { id });
  },

  async parseFile(filePath: string, parentId: number | null): Promise<ParsedCategories> {
    return invoke('parse_categories_file', { filePath, parentId });
  },

  async importFile(filePath: string, parentId: number | null): Promise<ImportSummary> {
    return invoke('import_categories_from_file', { filePath, parentId });
  },
};

File: os-compass/src-tauri/src/commands/category.rs
use crate::db::repository::CategoryRepository;
use crate::db::DbPool;
use crate::models::{Category, CreateCategoryInput};
use tauri::State;

#[tauri::command]
pub fn get_categories(pool: State<'_, DbPool>) -> Result<Vec<Category>, String> {
    let repo = CategoryRepository::new(&pool);
    repo.get_all()
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub fn get_category(
    pool: State<'_, DbPool>,
    id: i64,
) -> Result<Option<Category>, String> {
    let repo = CategoryRepository::new(&pool);
    repo.get_by_id(id)
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub fn create_category(
    pool: State<'_, DbPool>,
    input: CreateCategoryInput,
) -> Result<i64, String> {
    let repo = CategoryRepository::new(&pool);
    let category = Category {
        id: 0,
        name: input.name,
        path: input.path,
        explain: input.explain,
        sort_order: input.sort_order.unwrap_or(0),
        parent_id: input.parent_id,
        is_system: false,
        created_at: String::new(),
    };
    repo.create(&category)
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub fn update_category(
    pool: State<'_, DbPool>,
    category: Category,
) -> Result<(), String> {
    let repo = CategoryRepository::new(&pool);
    repo.update(&category)
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub fn delete_category(
    pool: State<'_, DbPool>,
    id: i64,
) -> Result<(), String> {
    let repo = CategoryRepository::new(&pool);
    repo.delete(id)
        .map_err(|e| e.to_string())
}
